src/resolve/resolved-dns-transaction.c

```diff
@@ -940,6 +940,34 @@ void dns_transaction_process_reply(DnsTransaction *t, DnsPacket *p) {
                         return;
                 }
 
+                /* Some captive portals are special in that the Aruba/Datavalet hardware will miss replacing the
+                 * packets with the local server IP to point to the authenticated side of the network if EDNS0 is
+                 * enabled. Instead they return NXDOMAIN, with DO bit set to zero... nothing to see here, yet respond
+                 * with the captive portal IP, when using the more simple UDP level.
+                 *
+                 * Common portal names that fail like so are:
+                 * secure.datavalet.io
+                 * securelogin.arubanetworks.com
+                 * securelogin.networks.mycompany.com
+                 *
+                 * Thus retry NXDOMAIN RCODES with a lower feature level.
+                 *
+                 * Do not "clamp" the feature level down, as the captive portal should not be lying for the wider
+                 * internet (e.g. _other_ queries were observed fine with EDNS0 on these networks, post auth)
+                 *
+                 * This is reported as https://github.com/dns-violations/dns-violations/blob/master/2018/DVE-2018-0001.md
+                 */
+                if (DNS_PACKET_RCODE(p) == DNS_RCODE_NXDOMAIN && t->current_feature_level >= DNS_SERVER_FEATURE_LEVEL_EDNS0) {
+                        char key_str[DNS_RESOURCE_KEY_STRING_MAX];
+                        dns_resource_key_to_string(t->key, key_str, sizeof key_str);
+                        t->current_feature_level = t->current_feature_level - 1;
```

Comment:
hmm, i think we should jump directly to DNS_SERVER_FEATURE_LEVEL_UDP here. I mean, why request with anything higher than one step below EDNS0 anyway, if we are not going to accept it?

Reply:
Ack.